### PR TITLE
Use the community instead of API Gateway to execute API calls

### PR DESCRIPTION
**Pull request by adityasharat**
* Refactor URL paths and response schema for all API calls.
* Add better error and debug logs.
* reuse `Gson` instance.
* Use `UriBuilder` instead of string concatenations.
* Improve unit test cases.
* Improve constant variable names for version.
* Improve `proguard` config.
* Refactor and deprecate the API Gateway Host in `LiAppCredentials`.
* Remove unnecessary throws clause in interface implementations.
* Check if SDK is initialized and a user is logged in before updating or saving device id for push notifications.
* Remove unnecessary use of primitive wrappers.

### CODE REVIEW

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/auth/LiAppCredentials.java [hunk 1]
```diff
@@ -136,9 +151,14 @@ public String getDeviceId() {
         return deviceId;
     }
 
+    /**
+     * The API Gateways host address.
+     *
+     * @return The API Gateway host Uri.
+     */
     @NonNull
     public Uri getApiGatewayHost() {
```
Comment (NarendraPrabhu) [inline li-android-sdk-core/src/main/java/lithium/community/android/sdk/auth/LiAppCredentials.java]: Can we get rid of this API getApiGatewayHost, or is it used at many places?

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/auth/LiAppCredentials.java [hunk 2]
```diff
@@ -297,7 +314,6 @@ public Builder setCommunityUri(@NonNull String communityUri) {
          */
         @NonNull
         public Builder setApiGatewayUri(@NonNull String apiGatewayUri) {
-            this.apiGatewayUri = apiGatewayUri;
```
Comment (NarendraPrabhu) [inline li-android-sdk-core/src/main/java/lithium/community/android/sdk/auth/LiAppCredentials.java]: same here, Can we get rid of this API setApiGatewayUri?

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiAuthManager.java [hunk 3]
```diff
@@ -244,15 +263,18 @@ public boolean isUserLoggedIn() {
...
                 return state;
             } catch (JSONException e) {
-                Log.e(LiAuthManager.class.getSimpleName(), "Auth State deserialization failed");
+                Log.e(LiCoreSDKConstants.LI_ERROR_LOG_TAG, "LiAuthManager#restoreAuthState() - deserialization of auth state failed");
                 e.printStackTrace();
             }
+        } else {
+            Log.d(LiCoreSDKConstants.LI_DEBUG_LOG_TAG, "LiAuthManager#restoreAuthState() - No saved auth state found");
         }
         return null;
```
Comment (NarendraPrabhu) [inline li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiAuthManager.java]: can we set state = null in else and catch blocks and have a single return statement, just for readability?

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiSDKManager.java [hunk 4]
```diff
@@ -159,68 +160,73 @@ private static void updatePreferences(@NonNull Context context) {
...
+                        e.printStackTrace();
+                    }
+                });
             } catch (LiRestResponseException e) {
-                Log.e(LOG_TAG, "ERROR: " + e);
+                Log.e(LiCoreSDKConstants.LI_ERROR_LOG_TAG, "User details request aborted");
+                e.printStackTrace();
             }
-
         }
```
Comment (NarendraPrabhu) [inline li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiSDKManager.java]: What if user is not logged in? what is expected as the output of this API? How are sync errors returned? two syncs in parallel? What if one fails and one succeeds?; Possibly it is a background job, Are there any other API calls dependent on any of these sync-ed settings? Can we have a callback mechanism?

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiSDKManager.java [hunk 5]
```diff
@@ -230,8 +236,20 @@ public void onError(Exception e) {
...
+     * Gets version name of the SDK.
+     *
+     * @return the version name of the SDK.
+     * @deprecated Use {@link #getVersion()} instead.
+     */
+    @Deprecated
+    @NonNull
     public String getCoreSDKVersion() {
-        return BuildConfig.li_sdk_core_version;
+        return getVersion();
```
Comment (NarendraPrabhu) [inline li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiSDKManager.java]: Can we get rid of this API?

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/auth/LiAppCredentials.java (hunk 1 above)
Comment (adityasharat) [reply li-android-sdk-core/src/main/java/lithium/community/android/sdk/auth/LiAppCredentials.java]: We cannot remove it, but this should have been marked deprecated. Will do that.

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/auth/LiAppCredentials.java (hunk 2 above)
Comment (adityasharat) [reply li-android-sdk-core/src/main/java/lithium/community/android/sdk/auth/LiAppCredentials.java]: For maintaining backward compatibility we cannot remove it. The `Builder` class is entirely deprecated.

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiAuthManager.java (hunk 3 above)
Comment (adityasharat) [reply li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiAuthManager.java]: Will do that.

File: li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiSDKManager.java (hunk 4 above)
Comment (adityasharat) [reply li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiSDKManager.java]: Eventually, we should broadcast an intent after the sync tasks are completed.

Right Now the API returns void; both the API calls are independent of each other and async. If the user is not logged in then this is a No-Op.

Comment (NarendraPrabhu) [reply li-android-sdk-core/src/main/java/lithium/community/android/sdk/manager/LiSDKManager.java]: Perhaps! an async-callback mechanism. And probably going further we can separate out the two APIs.

**Comment by NarendraPrabhu** [comment]
Reviewed.

**Pull request merged**